### PR TITLE
Add documentation on kind signatures, which ones are displayed, and why

### LINKED ISSUE #434: Add an explanation on kind signatures appearing in documentation
Follow up from this comment: https://github.com/purescript/purescript/pull/4119#issuecomment-869022915

Below is what I originally wrote for the changelog entry:

---

Previously, data/newtype/type/class declarations that have explicit kind
signatures would not display those kind signatures in their documentation.
For example, the two below types...

```purescript
data PolyProxy :: forall k. k -> Type
data PolyProxy a = PolyProxy

data TypeProxy :: Type -> Type
data TypeProxy a = TypeProxy
```

... would only show the following information in their docs. One cannot
be distinguished from another due to the missing kind signatures:

```
data PolyProxy a = PolyProxy

data TypeProxy a = TypeProxy
```

Now, these types' kind signatures are displayed above their declarations
in their docs, similar to what one would see in the source code. The only
exception to this are kind signatures whose kinds are "uninteresting"
(covered below).

Moreover, declarations without explicit kind signatures will still
display the kind signatures inferred by the compiler. For example,
`Foo`'s documentation will display the below (commented-out) inferred
kind signature:

```purescript
                                                {-
data Foo :: (Type -> Type) -> Type -> Type      -}
data Foo f a = Foo (f Int) a
```

However, a kind signature for both explicit and inferred kinds will not
be displayed if that kind is considered "uninteresting." An "uninteresting"
kind is one that follows this form:
- `Type`
- `Constraint`
- `Type -> K` where `K` refers to an "uninteresting" kind signature

Thus, both the `Bar` and `Baz` declarations below will not have their
explicit and inferred kind signatures displayed in their docs.
```purescript
data Bar :: Type -> Type -> Type
data Bar a b = Bar a b
                                                  {-
class Baz :: Type -> Type -> Constraint           -}
class Baz a b where
  doBaz :: a -> b -> String
```

### PR DESCRIPTION
**Pull request by JordanMartinez**
Fixes #434 

I'll have to update the changelog with an entry for this after #436 gets merged.

### CODE REVIEW

File: static/help-docs/users.md [hunk 1]
```diff
@@ -79,3 +79,125 @@ called "Kleisli" in some contexts. This may also change in the future; see
...
+```purescript
+-- Explicit kind signature
+
+data ExplicitFoo :: forall k. k -> Type
+data ExplicitFoo a = ExplicitFoo
+
+-- Implicit kind signature
+                                                                  {-
+data ImplicitFoo :: forall k. k -> (Type -> Type) -> Type         -}
+data ImplicitFoo a f = ImplicitFoo (f Int)
```
Comment (thomashoneyman) [inline static/help-docs/users.md:100]: Perhaps we could do something to make it a little clearer the signature is being inferred by something like:

```purs
{- The compiler will infer this kind signature for the below type:
data ImplicitFoo :: forall k. k -> (Type -> Type) -> Type
-}
data ImplicitFoo a f = ImplicitFoo (f Int)

File: static/help-docs/users.md [hunk 2]
```diff
@@ -79,3 +79,125 @@ called "Kleisli" in some contexts. This may also change in the future; see
...
+-- | Data declaration documentation comment
+data ExplicitFoo a = ExplicitFoo
+```
+... will be rendered as
+```
+Kind signature declaration documentation comment
+Data declaration documentation comment
+```
+
+### Interesting Kinds are displayed; Uninteresting Kinds are not
```
Comment (thomashoneyman) [inline static/help-docs/users.md:122]: I don't think "Kind" has to be capitalized

```suggestion
### Interesting kinds are displayed; uninteresting kinds are not
```

File: static/help-docs/users.md [hunk 3]
```diff
@@ -79,3 +79,125 @@ called "Kleisli" in some contexts. This may also change in the future; see
...
+if the kind signatures are considered "interesting." Put differently,
+"uninteresting" kind signatures will not be displayed.
+
+An "uninteresting" kind signature is one that follows this form:
+- `Type`
+- `Constraint`
+- `Type -> K` where `K` refers to an "uninteresting" kind signature
+
+Here's another way to think about it. Kind signatures are considered
+"uninteresting" if all of their type parameters' kinds are kind `Type`.
```
Comment (thomashoneyman) [inline static/help-docs/users.md:140]:
```suggestion
Here's another way to think about it: kind signatures are considered
"uninteresting" if all of their type parameters' kinds are of kind `Type`.
```

File: static/help-docs/users.md [hunk 4]
```diff
@@ -79,3 +79,125 @@ called "Kleisli" in some contexts. This may also change in the future; see
...
+
+-- every type parameter makes this kind signature "interesting"
+class TypeLevelProgrammingFunction :: Symbol -> Row Type -> Row Type -> Constraint
+class TypeLevelProgrammingFunction sym row1 row2 | sym row1 -> row2
+```
+
+In the `PolyProxy` example, the polykinded `a` type parameter is interesting.
+
+In the `FunctorLike` and `BiFunctorLike` examples, the higher-kinded `f`
+type paramter is interesting.
```
Comment (thomashoneyman) [inline static/help-docs/users.md:201]:
```suggestion
In the `FunctorLike` and `BiFunctorLike` examples, the higher-kinded `f`
type parameter is interesting.
```

**Review by thomashoneyman — COMMENTED** [review]
I've included a few comments. In general this feels a little long; I wonder if we could trim some of the details or move them into the `purs docs` documentation instead.

I'm also not totally sold on the 'interesting' vs. 'uninteresting' terminology, though I don't have a better pairing in mind at the moment.

**Comment by JordanMartinez** [comment]
I've addressed all feedback. I tried making it shorter by removing a few things. How is it now?

**Review by kl0tl — COMMENTED** [review]
The example sections feel redundant with the “Interesting kinds are displayed; Uninteresting kinds are not” section. Perhaps we could provide examples when explaining what kinds are uninteresting and then show some interesting kinds as counterexamples?

**Comment by JordanMartinez** [comment]
> Perhaps we could provide examples when explaining what kinds are uninteresting and then show some interesting kinds as counterexamples?

Isn't that what I already do in this PR in the subsections?

File: static/help-docs/users.md (hunk 1 above)
Comment (JordanMartinez) [reply static/help-docs/users.md]: @kl0tl Here's what I was doing above originally. Thomas is the one who suggested the edit that you commented on below.

Which is preferred?

**Comment by kl0tl** [comment]
Yes, but I don’t think it is necessary to rehash what uninteresting kinds are in the “Examples of "uninteresting" kind signatures“ section.

**Comment by JordanMartinez** [comment]
How are the latest changes?

**Review by thomashoneyman — APPROVED** [review]

**Comment by JordanMartinez** [comment]
@thomashoneyman Thanks for the approval! This has been merged. We can redeploy Pursuit and finally announce v0.14.3

**Pull request merged**